boilermaker/app.py

```diff
@@ -106,13 +115,28 @@ async def publish_task(self, task: Task, delay: int = 0, retries: int = 3):
                 ServiceBusAuthorizationError,
                 ServiceBusAuthenticationError,
             ) as e:
-                encountered_errors.append(e)    
+                encountered_errors.append(e)
         else:
             raise BoilermakerAppException(
                 "Error encountered while publishing task to service bus",
                 encountered_errors,
             )
 
+    async def signal_handler(self, receiver: ServiceBusReceiver, scope: CancelScope):
+        """We would like to reschedule any open messages on SIGINT/SIGTERM"""
+        with open_signal_receiver(signal.SIGINT, signal.SIGTERM) as signals:
+            async for signum in signals:
+                if self._current_message is not None:
+                    sequence_number = self._current_message.sequence_number
+                    await receiver.abandon_message(self._current_message)
```

Comment:
Dumb question, but do we need anything to handle if the message is already settled? If I understand correctly; if we try to abandon something that managed to process already, we will error out here.

Reply:
Actually, this could probably be ignored. Worst case is we get an extra log line, but the message will be taken care of regardless.

Reply:
Not a dumb question and it's worth thinking through. Consider the following flow:
1. message work completed
1. message marked `completed` (`await` point)
1. SIGTERM -> `abandon_message`   # this should error out
1. This would be when _other_ async task tries to set `self._current_message = None`

At the very least, I will probably need to catch errors on `abandon_message` 🤔

Reply:
The following exceptions that [can be raised when calling this method](https://github.com/Azure/azure-sdk-for-python/blob/75d6c26c724f820b93d818f8ed5d91efba7d4d30/sdk/servicebus/azure-servicebus/azure/servicebus/aio/_servicebus_receiver_async.py#L851):

```
        :raises: ~azure.servicebus.exceptions.MessageAlreadySettled if the message has been settled.
        :raises: ~azure.servicebus.exceptions.SessionLockLostError if session lock has already expired.
        :raises: ~azure.servicebus.exceptions.ServiceBusError when errors happen.
```

I'm going to catch these and move on for now.

Reply:
> Actually, can just be ignored. Worst case is we get an extra log line, but the message will be taken care of regardless.

No, you were right. It's dangerous not to catch these exceptions: it would crash the whole client and prevent it from making progress.